packages/xo-web/src/xo-app/home/index.js

```diff
@@ -700,21 +704,29 @@ export default class Home extends Component {
     return event => setFilter(event.target.value)
   })
 
+  _getEnabledJobs = createSelector(
+    () => this.props.jobs,
+    () => this.props.schedulesByJob,
+    createCollectionWrapper((jobs, schedulesByJob) => filter(jobs, job => some(schedulesByJob[job.id], 'enabled')))
+  )
+
+  _getFilterBackedUpVms = createSelector(
+    this._getEnabledJobs,
+    enabledJobs => vms => uniq(flatMap(enabledJobs, job => filter(vms, createPredicate(omit(job.vms, 'power_state')))))
+  )
+
   _getFilteredItems = createSort(
     createSelector(
       createFilter(() => this.props.items, this._getFilterFunction),
       () => this.props.location.query.backup,
-      () => this.props.jobs,
-      (filteredItems, backup, jobs) => {
+      this._getEnabledJobs,
+      this._getFilterBackedUpVms,
+      (filteredVms, backup, enabledJobs, filterBackedUpVms) => {
         if (backup === undefined) {
-          return filteredItems
+          return filteredVms
         }
-
-        const backedUpVms = uniq(
-          flatMap(jobs, job => filter(filteredItems, createPredicate(omit(job.vms, 'power_state'))))
-        )
-
-        return backup === 'true' ? backedUpVms : differenceBy(map(filteredItems), backedUpVms, 'id')
+        const backedUpVms = filterBackedUpVms(filteredVms)
+        return backup === 'true' ? backedUpVms : differenceBy(map(filteredVms), backedUpVms, 'id')
```

Comment:
Alternate proposal:

```js
_getFilteredItems = createSort(
  createSelector(
    createFilter(() => this.props.items, this._getFilterFunction),
    createSelector(
      () => this.props.location.query.backup,
      () => this.props.jobs,
      () => this.props.schedulesByJob,
      (backup, jobs, schedulesByJob) => {
        if (backup !== undefined) {
          const pattern = {
            __or: jobs
              .filter(job => schedulesByJob[job.id].some(_ => _.enabled))
              .map(job => omit(job.vms, 'power_state')),
          }
          return createPredicate(backup === 'true' ? pattern : { __not: pattern })
        }
      }
    ),
    (filteredVms, predicate) => (predicate === undefined ? filteredVms : filter(filteredVms, predicate))
  ),
  createSelector(
    () => this.state.sortBy,
    sortBy => [sortBy, 'name_label']
  ),
  () => this.state.sortOrder
)
```

Differences with your  code:
- don't use `createCollectionWrapper`, it's not really necessary, `jobs` and `schedules` don't change often
- don't create the predicate if `backup` is `undefined` (which is almost always the case)
- only use a single predicate thanks to `{ __or: Pattern[] }` (so don't need multiple passes onf VMs and `uniq` or `flatMap`)
- don't need to compute the difference thanks to `{ __not: Pattern }`
- while we're at it, use native methods instead of Lodash

What's your opinion on this?

Reply:
Just so I understand:
- using `createCollectionWrapper` when the input selectors don't change often isn't bad, right? I guess you meant that this won't make your implementation less performant because of this?
- the previous implementation did create the predicate but only once (and then when the jobs change, which doesn't happen often)
- agreed
- agreed too
- why not :)

Reply:
- yes using `createCollectionWrapper` is not a problem, but I'm not sure it provides any real perf improvement here, in this case, I think we should go for simplicity and remove it :)
- you're right, it's not a big deal but I think it's almost always unnecessary so why not avoid it entirely

Using selectors like this makes me realize how much lazy selectors like the ones provided by Reaclette are powerful: we don't have to think about optimization, the inputs are only computed when necessary, making the code much more natural.